### PR TITLE
SDSTOR-11060 : range_put bug due to lack of available space in a node

### PR DESCRIPTION
**Pull request by shosseinimotlagh**
Before update the leaf node, a check is added to make sure there is enough available size for new value

### CODE REVIEW

**Comment by yamingk** [comment]
Should we target the bug fix (unrelated to fiber) to master branch? The master branch need this fix to unblock other bug fixes (I have one pending)?

File: src/include/homestore/btree/detail/btree_mutate_impl.ipp [hunk 1]
```diff
@@ -203,7 +203,12 @@ btree_status_t Btree< K, V >::mutate_write_leaf_node(const BtreeNodePtr& my_node
                 BT_DBG_ASSERT(false, "For non-extent keys range-update should be really update and cannot insert");
                 ret = btree_status_t::not_supported;
             } else {
+                const auto new_val_size{(*req.m_newval).serialized_size()};
                 for (auto idx{start_idx}; idx <= end_idx; ++idx) {
+                    if (my_node->available_size(m_bt_cfg) <= new_val_size) {
```
Comment (yamingk) [inline src/include/homestore/btree/detail/btree_mutate_impl.ipp:208]: I don't quite follow what is this trying to achieve, but result of this line: 208 doesn't seem to rely on the for loop. 
If it is false, it is always false, if it is true, it always do with the 1st idx. 

Also shouldn't the has_more be set to ret and fall through to below logic to handle it? I don't see any caller is handling has_more.

Comment (shosseinimotlagh) [reply src/include/homestore/btree/detail/btree_mutate_impl.ipp:208]: 1- `update()` changes the value of available_size. Let's assume the current val for entry _`idx`_  has size 1 and the the requested val size = 500. Therefore, it needs to check the requested val size is available or not. (There is a minor bug that I have fixed it to check  extra needed size which I will update this PR). 

2- For the second point this function should return `has_more` then in `put()` it checks the return value and when it is `has_more`, it calls `do_put()` again and in the next call of `do_put()` , `split_nodes()` will occur and the rest of request will perfom.

File: src/include/homestore/btree/detail/btree_mutate_impl.ipp [hunk 2]
```diff
@@ -203,7 +203,12 @@ btree_status_t Btree< K, V >::mutate_write_leaf_node(const BtreeNodePtr& my_node
...
+                        req.set_cursor_key(my_node->get_nth_key< K >(idx, true));
```
Comment (yamingk) [inline src/include/homestore/btree/detail/btree_mutate_impl.ipp:209]: Is there a reason to copy the key?

File: src/include/homestore/btree/detail/btree_mutate_impl.ipp [hunk 3]
```diff
@@ -203,7 +203,12 @@ btree_status_t Btree< K, V >::mutate_write_leaf_node(const BtreeNodePtr& my_node
                 BT_DBG_ASSERT(false, "For non-extent keys range-update should be really update and cannot insert");
                 ret = btree_status_t::not_supported;
             } else {
+                const auto new_val_size{(*req.m_newval).serialized_size()};
```
Comment (yamingk) [inline src/include/homestore/btree/detail/btree_mutate_impl.ipp:206]: What will happen if start_found or end_found is not true?

Comment (shosseinimotlagh) [reply src/include/homestore/btree/detail/btree_mutate_impl.ipp:206]: When we are in this point, we are sure that we are in the leaf. Before calling this function the request is getting trimmed `req.trim_working_range( my_node->min_of(` for the leaf. 
For the out of band case, if we point to same start and end without any match, it is hitting unavailable range. In this case, `count = my_node->template get_all< K, V > `returns 0 and `BT_NODE_LOG_ASSERT` would be raised accordingly.

File: src/include/homestore/btree/detail/btree_mutate_impl.ipp (hunk 2 above)
Comment (shosseinimotlagh) [reply src/include/homestore/btree/detail/btree_mutate_impl.ipp:209]: confirmed. no need

File: src/include/homestore/btree/detail/btree_mutate_impl.ipp (hunk 3 above)
Comment (yamingk) [reply src/include/homestore/btree/detail/btree_mutate_impl.ipp:206]: Can we add below test case with rangeput, the expected result should be RangePut should fail. 
Case-1:  start_key is invalid, end_key is valid.
Case-2: start_key is valid, end_key is invalid.
Case-3: both keys are invalid.

File: src/include/homestore/btree/detail/btree_mutate_impl.ipp [hunk 4]
```diff
@@ -203,7 +203,14 @@ btree_status_t Btree< K, V >::mutate_write_leaf_node(const BtreeNodePtr& my_node
...
+                V tmp_v;
                 for (auto idx{start_idx}; idx <= end_idx; ++idx) {
+                    my_node->get_nth_value(idx, &tmp_v, false);
+                    if (my_node->available_size(m_bt_cfg) + tmp_v.serialized_size() <= new_val_size) {
+                        req.set_cursor_key(my_node->get_nth_key< K >(idx, false));
```
Comment (yamingk) [inline src/include/homestore/btree/detail/btree_mutate_impl.ipp:211]: What is the reason we don't update req.working_range here?

**Pull request closed without merging**